docs/modules.md

```diff
@@ -181,13 +181,68 @@ The arguments passed to this callback are:
 ```python
 async def check_media_file_for_spam(
     file_wrapper: "synapse.rest.media.v1.media_storage.ReadableFileWrapper",
-    file_info: "synapse.rest.media.v1._base.FileInfo"
+    file_info: "synapse.rest.media.v1._base.FileInfo",
 ) -> bool
 ```
 
 Called when storing a local or remote file. The module must return a boolean indicating
 whether the given file can be stored in the homeserver's media store.
 
+#### Third party rules callbacks
+
+Third party rules callbacks allow modules developers to add extra checks to verify the
+validity of incoming events. Third party event rules callbacks can be registered using
+the module API's `register_third_party_rules_callbacks` method.
+
+The available third party rules callbacks are:
+
+```python
+async def check_event_allowed(
+    event: "synapse.events.EventBase",
+    state_events: "synapse.types.StateMap",
+) -> Tuple[bool, Optional[dict]]
+```
+
+Called when processing any incoming event, with the event and the a `StateMap`
+representing the current state of the room the event is being sent into. A `StateMap` is
+a dictionary indexed on tuples containing an event type and a state key; for example
+retrieving the room's `m.room.create` event from the `state_events` argument looks like
+this: `state_events.get(("m.room.create", ""))`. The module must return a boolean
+indicating whether the event can be allowed.
+
+Note that this callback function processes incoming events coming via federation
+traffic (on top of client traffic). This means denying an event might cause the local
+copy of the room's history to diverge from the ones of remote servers. This may cause
+federation issues in the room. It is strongly recommended to only deny events using this
+callback function if the sender is a local user, or in a private federation in which all
+servers are using the same module, with the same configuration.
```

Comment:
Hmm, why do we even have this functionality then?

Additionally, it might be worth calling out that the spam checker callback may be better suited for handling unwanted events over federation.

Reply:
> Hmm, why do we even have this functionality then?

Because it's useful to have in private federations, which is what the doc says.

> Additionally, it might be worth calling out that the spam checker callback may be better suited for handling unwanted events over federation.

Yeah good idea. There are some other issues with this method which would justify even more recommending using the spam checker callback instead.